src/Data/Singletons/TH/Options.hs

```diff
@@ -0,0 +1,255 @@
+{-# LANGUAGE GeneralizedNewtypeDeriving #-}
+{-# LANGUAGE TemplateHaskell #-}
+
+-----------------------------------------------------------------------------
+-- |
+-- Module      :  Data.Singletons.TH.Options
+-- Copyright   :  (C) 2019 Ryan Scott
+-- License     :  BSD-style (see LICENSE)
+-- Maintainer  :  Ryan Scott
+-- Stability   :  experimental
+-- Portability :  non-portable
+--
+-- This module defines 'Options' that control finer details of how the Template
+-- Haskell machinery works, as well as an @mtl@-like 'OptionsMonad' class
+-- and an 'OptionsM' monad transformer.
+--
+----------------------------------------------------------------------------
+
+module Data.Singletons.TH.Options
+  ( -- * Options
+    Options, defaultOptions
+    -- ** Options record selectors
+  , genSingKindInsts
+  , promotedClassName
+  , promotedValueName
+  , singledDataTypeName
+  , singledClassName
+  , singledDataConName
+  , singledValueName
+  , defunctionalizedName
+    -- ** Derived functions over Options
+  , promotedTopLevelValueName
+  , promotedLetBoundValueName
+  , defunctionalizedName0
+
+    -- * OptionsMonad
+  , OptionsMonad(..), OptionsM, withOptions
+  ) where
+
+import Control.Applicative
+import Control.Monad.IO.Class (MonadIO)
+import Control.Monad.Reader (ReaderT(..), ask)
+import Control.Monad.RWS (RWST)
+import Control.Monad.State (StateT)
+import Control.Monad.Trans.Class (MonadTrans(..))
+import Control.Monad.Writer (WriterT)
+import Data.Singletons.Names ( consName, listName, nilName
+                             , mk_name_tc, mkTupleDataName, mkTupleTypeName
+                             , sconsName, sListName, snilName
+                             , splitUnderscores
+                             )
+import Data.Singletons.Util
+import Language.Haskell.TH.Desugar
+import Language.Haskell.TH.Syntax hiding (Lift(..))
+
+-- | Options that control the finer details of how @singletons@' Template
+-- Haskell machinery works.
+data Options = Options
+  { genSingKindInsts        :: Bool
+    -- ^ If 'True', then 'SingKind' instances will be generated. If 'False',
+    --   they will be omitted entirely. This can be useful in scenarios where
+    --   TH-generated 'SingKind' instances do not typecheck (for instance,
+    --   when generating singletons for GADTs).
+  , promotedClassName    :: Name -> Name
+    -- ^ Given the name of the original, unrefined class, produces the name of
+    --   the promoted equivalent of the class.
+  , promotedValueName    :: Name -> Maybe Uniq -> Name
```

Comment:
I like this design *much* more than the previous one. I haven't looked for knock-on effects, but if they are reasonable, then I think this is a nice way to do what you want.

Reply:
> I haven't looked for knock-on effects, but if they are reasonable, then I think this is a nice way to do what you want.

There are no knock-on _downsides_ to the best of my knowledge. There may be a knock-on _benefit_ in that this patch fixes a subtle bug related to how local fixity declarations are promoted. The bugfix is contained in this part of the diff:

```diff
@@ -615,20 +629,19 @@ promoteInfixDecl name fixity = do
     -- Certain function names do not change when promoted (e.g., infix names),
     -- so don't bother with them.
     promote_infix_val :: q (Maybe DDec)
-    promote_infix_val
-      | nameBase name == nameBase promoted_name
-      = never_mind
-      | otherwise
-      = finish promoted_name
-
-    promoted_name :: Name
-    promoted_name = promoteValNameLhs name
+    promote_infix_val = do
+      opts <- getOptions
+      let promoted_name :: Name
+          promoted_name = promotedValueName opts name mb_let_uniq
+      if nameBase name == nameBase promoted_name
+         then never_mind
+         else finish promoted_name
```

Notice that the old code used `promoteValNameLhs` (the previous name for `promotedTopLevelName`), but this is just plain wrong in the presence of local fixity declarations like this one:

```hs
f :: Bool
f = g () () where
  infixl 4 `g`
  g :: () -> () -> Bool
  g _ _ = True
```

Interestingly, this bug can't be observed in today's `singletons` due to [this GHC bug](https://gitlab.haskell.org/ghc/ghc/issues/17608) that causes all TH-quoted local fixity declarations to be dropped. Once that bug is fixed (which [should be soon](https://gitlab.haskell.org/ghc/ghc/merge_requests/2372)), then the code above would error if singled. However, this patch solves the problem by taking the `let`-bound unique (`mb_let_uniq`) into account when promoting the names in a fixity declaration.